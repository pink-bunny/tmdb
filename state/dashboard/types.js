export const FETCH_TRENDING_MOVIES = 'FETCH_TRENDING_MOVIES';
export const FETCH_TRENDING_MOVIES_SUCCESS = 'FETCH_TRENDING_MOVIES_SUCCESS';
export const FETCH_TRENDING_MOVIES_ERROR = 'FETCH_TRENDING_MOVIES_ERROR';
