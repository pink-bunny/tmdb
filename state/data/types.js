export const MERGE_MOVIES_LIST = 'MERGE_MOVIES_LIST';

export const MERGE_LISTS_LIST = 'MERGE_LISTS_LIST';
