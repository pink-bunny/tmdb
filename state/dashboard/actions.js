import * as types from './types';

export const fetchTrendingMovies = (page = 1) => ({
  type: types.FETCH_TRENDING_MOVIES,
  page
});

export const fetchTrendingMoviesSuccess = (ids, totalItems, currentPage) => ({
  type: types.FETCH_TRENDING_MOVIES_SUCCESS,
  ids,
  totalItems,
  currentPage
});

export const fetchTrendingMoviesError = (error) => ({
  type: types.FETCH_TRENDING_MOVIES_ERROR,
  error
});

export const searchMovies = ({ search }, setErrors, setSubmitting, page = 1) => ({
  type: types.SEARCH_MOVIES,
  search,
  page,
  setErrors,
  setSubmitting
});
