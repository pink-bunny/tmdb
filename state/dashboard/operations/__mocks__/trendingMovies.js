/* eslint-disable quote-props, quotes */
export const trendingMoviesResponseSuccess = {
  "data": {
    "page": 1,
    "results": [
      {
        "id": 512200,
        "video": false,
        "vote_count": 1938,
        "vote_average": 6.8,
        "title": "Jumanji: The Next Level",
        "release_date": "2019-12-04",
        "original_language": "en",
        "original_title": "Jumanji: The Next Level",
        "genre_ids": [
          28,
          12,
          35,
          14
        ],
        "backdrop_path": "/hreiLoPysWG79TsyQgMzFKaOTF5.jpg",
        "adult": false,
        "overview": "As the gang return to Jumanji to rescue one of their own, they discover that nothing is as they expect. The players will have to brave parts unknown and unexplored in order to escape the worldâ€™s most dangerous game.",
        "poster_path": "/bB42KDdfWkOvmzmYkmK58ZlCa9P.jpg",
        "popularity": 267.705,
        "media_type": "movie"
      },
      {
        "id": 330457,
        "video": false,
        "vote_count": 2626,
        "vote_average": 7.1,
        "title": "Frozen II",
        "release_date": "2019-11-20",
        "original_language": "en",
        "original_title": "Frozen II",
        "genre_ids": [
          12,
          16,
          10751
        ],
        "backdrop_path": "/xJWPZIYOEFIjZpBL7SVBGnzRYXp.jpg",
        "adult": false,
        "overview": "Elsa, Anna, Kristoff and Olaf head far into the forest to learn the truth about an ancient mystery of their kingdom.",
        "poster_path": "/pjeMs3yqRmFL3giJy4PMXWZTTPa.jpg",
        "popularity": 145.51,
        "media_type": "movie"
      }
    ],
    "total_pages": 1000,
    "total_results": 20000
  }
};

export const trendingMoviesResponseError = {
  "response": {
    "data": {
      "status_code": 22,
      "status_message": "Invalid page: Pages start at 1 and max at 1000. They are expected to be an integer."
    }
  }
};
