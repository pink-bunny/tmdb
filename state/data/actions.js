import * as types from './types';
import { TOGGLE_TO_WATCHLIST_SUCCESS } from '../movie/types';

export const mergeMoviesList = (movies) => ({
  type: types.MERGE_MOVIES_LIST,
  movies
});

export const mergeListsList = (lists) => ({
  type: types.MERGE_LISTS_LIST,
  lists
});

export const toggleToWatchlistSuccess = (id, isMovieInWatchlist) => ({
  type: TOGGLE_TO_WATCHLIST_SUCCESS,
  id,
  isMovieInWatchlist
});
