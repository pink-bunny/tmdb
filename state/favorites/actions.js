import * as types from './types';

export const fetchFavorites = (page = 1) => ({
  type: types.FETCH_FAVORITES,
  page
});
export const fetchFavoritesSuccess = (ids, totalItems, currentPage) => ({
  type: types.FETCH_FAVORITES_SUCCESS,
  ids,
  totalItems,
  currentPage
});

export const fetchFavoritesError = (error) => ({
  type: types.FETCH_FAVORITES_ERROR,
  error
});

export const toggleToFavorite = (id, favorite, needRefetchList = false) => ({
  type: types.TOGGLE_TO_FAVORITES,
  id,
  favorite,
  needRefetchList
});

export const toggleToFavoriteSuccess = (id, isMovieInFavorites) => ({
  type: types.TOGGLE_TO_FAVORITES_SUCCESS,
  id,
  isMovieInFavorites
});
