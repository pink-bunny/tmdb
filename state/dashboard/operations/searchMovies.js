import { createLogic } from 'redux-logic';

import normalizedMovies from '../../../utils/normalize/movies';
import { SEARCH_MOVIES } from '../types';
import { mergeMoviesList } from '../../data/actions';
import {
  fetchMoviesSuccess,
  fetchMoviesError
} from '../actions';

const searchMoviesLogic = createLogic({
  type: SEARCH_MOVIES,

  async process({ httpClient, action }, dispatch, done) {
    const { searchQuery, page } = action;
    const currentPage = page;

    try {
      const { data } = await httpClient.get(`search/movie?query=${searchQuery}&page=${page}`);
      // normalized data
      const { movies, ids, totalItems } = normalizedMovies(data);

      dispatch(mergeMoviesList(movies));
      dispatch(fetchMoviesSuccess(ids, totalItems, currentPage));
    } catch (error) {
      const [errorMessage] = error.response.data.errors;
      dispatch(fetchMoviesError(errorMessage));
    }
    done();
  }
});

export default searchMoviesLogic;
