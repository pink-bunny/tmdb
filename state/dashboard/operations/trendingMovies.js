import { createLogic } from 'redux-logic';

import normalizedMovies from '../../../utils/normalize/movies';
import { FETCH_TRENDING_MOVIES } from '../types';
import { fetchMoviesSuccess, fetchMoviesError } from '../actions';
import { mergeMoviesList } from '../../data/actions';

const trendingMoviesLogic = createLogic({
  type: FETCH_TRENDING_MOVIES,

  async process({ httpClient, action }, dispatch, done) {
    const { page } = action;
    const currentPage = page;
    try {
      const { data } = await httpClient.get(`trending/movie/day?page=${page}`);

      // normalized data
      const { movies, ids, totalItems } = normalizedMovies(data);

      dispatch(mergeMoviesList(movies));
      dispatch(fetchMoviesSuccess(ids, totalItems, currentPage));
    } catch (error) {
      const errorMessage = error.response.data.status_message;
      dispatch(fetchMoviesError(errorMessage));
    }
    done();
  }
});

export default trendingMoviesLogic;
