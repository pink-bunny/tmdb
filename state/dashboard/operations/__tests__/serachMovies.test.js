import searchMovies from '../searchMovies';
import { SEARCH_MOVIES } from '../../types';
import mockHttpClient from '../../../../utils/testsHelpers/mockHttpClient';
import { moviesResponseSuccess, searchMoviesResponseError } from '../__mocks__/trendingMovies';
import normalizedMovies from '../../../../utils/normalize/movies';
import { mergeMoviesList } from '../../../data/actions';
import {
  fetchMoviesSuccess,
  fetchMoviesError
} from '../../actions';

jest.mock('../../../data/actions', () => ({
  mergeMoviesList: jest.fn()
}));
jest.mock('../../actions', () => ({
  fetchMoviesSuccess: jest.fn(),
  fetchMoviesError: jest.fn()
}));

describe('searchMovies logic', () => {
  const action = {
    type: SEARCH_MOVIES,
    page: 1,
    searchQuery: 'query'
  };
  const dispatch = jest.fn();
  const done = jest.fn();
  const { searchQuery, page } = action;
  const currentPage = page;
  let httpClient;
  httpClient = mockHttpClient({ method: 'get', response: moviesResponseSuccess });
  searchMovies.process({ httpClient, action }, dispatch, done);

  it('matches snapshot', () => {
    expect(searchMovies).toMatchSnapshot();
  });

  it('httpClient called with right arguments', () => {
    const requestUrl = `search/movie?query=${searchQuery}&page=${page}`;
    expect(httpClient.get).toHaveBeenCalledWith(requestUrl);
  });

  describe('success', () => {
    const { data } = moviesResponseSuccess;
    const { movies, ids, totalItems } = normalizedMovies(data);

    it('dispatch mergeMoviesList() with movies', () => {
      expect(mergeMoviesList).toHaveBeenCalledWith(movies);
    });
    it('dispatch fetchMoviesSuccess() with ids, totalItems, currentPage', () => {
      expect(fetchMoviesSuccess).toHaveBeenCalledWith(ids, totalItems, currentPage);
    });
  });

  describe('failure', () => {
    httpClient = mockHttpClient({ method: 'get', response: searchMoviesResponseError, reject: true });
    searchMovies.process({ httpClient, action }, dispatch, done);

    it('dispatch fetchMoviesError() with error', () => {
      const error = searchMoviesResponseError;
      const [errorMessage] = error.response.data.errors;
      expect(fetchMoviesError).toHaveBeenCalledWith(errorMessage);
    });
  });

  it('done() have been called', () => {
    expect(done).toHaveBeenCalled();
  });
});
