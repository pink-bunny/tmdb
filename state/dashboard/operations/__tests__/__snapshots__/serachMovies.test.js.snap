// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchMovies logic matches snapshot 1`] = `
Object {
  "cancelType": undefined,
  "debounce": 0,
  "latest": false,
  "name": undefined,
  "process": [Function],
  "processOptions": Object {
    "dispatchMultiple": true,
  },
  "throttle": 0,
  "transform": undefined,
  "type": "SEARCH_MOVIES",
  "validate": undefined,
  "warnTimeout": 60000,
}
`;
